streams/jvm/src/main/scala/zio/stream/compression/Gunzipper.scala

```diff
@@ -0,0 +1,181 @@
+package zio.stream.compression
+
+import java.util.zip.{ CRC32, Inflater }
+import java.{ util => ju }
+
+import scala.annotation.tailrec
+
+import zio._
+
+/** Performs few steps of parsing header, then decompresses body and checks trailer.
+ * With reasonably chosen bufferSize there shouldn't occur many concatenation of arrays. */
+private[compression] class Gunzipper private (bufferSize: Int) {
+
+  import Gunzipper._
+
+  private var state: State = new ParseHeaderStep(Array.emptyByteArray, new CRC32)
+
+  def reset(): Unit = {
+    state.close()
+    state = new ParseHeaderStep(Array.emptyByteArray, new CRC32)
+  }
+
+  def close(): Unit = state.close()
+
+  def onChunk(c: Chunk[Byte]): ZIO[Any, CompressionException, Chunk[Byte]] =
```

Comment:
In terms of code organization, I'd be fine with this just being `: Chunk[Byte]` and wrapping the exceptions into the `CompressionException` in `compression.gunzip`.